### PR TITLE
[GLib] Need API for asynchronously handling WebKitDownload::decide-destination

### PR DESCRIPTION
**Pull request by mcatanzaro**
#### 878af52cf35811d107d6bb6f38531ce3b73f211d
<pre>
[GLib] Need API for asynchronously handling WebKitDownload::decide-destination
<a href="https://bugs.webkit.org/show_bug.cgi?id=238748">https://bugs.webkit.org/show_bug.cgi?id=238748</a>

Reviewed by Carlos Garcia Campos.

Currently WebKitDownload::decide-destination is emitted when WebKit
wants to know the filesystem location to download the file to. The
application is expected to display a file chooser using
gtk_dialog_run(), then return the result to WebKit using
webkit_download_set_destination(). In GTK 4, gtk_dialog_run() no longer
exists. The only way to accomplish this is to hande it manually,
iterating the default GMainContext inside the signal handler callback.
This is called a &quot;nested main loop&quot; and it&apos;s very dangerous because it
frequently results in code being executed at unexpected times. For
example, the attached bug report shows an unintended but also
unavoidable example of a triple nested main loop that occurs where
a second download starts inside the first decide-destination signal
handler, then another, then another. Not good. (Note this is not a new
problem in GTK 4. gtk_dialog_run() no longer exists because it
internally ran a nested main loop, which had this exact same problem.)

To fix this, we need to allow applications to handle the
decide-destination signal asynchronously, so they can finish the signal
handler without immediately providing a destination, provide it later on
in a future main context iteration, and know that the download will not
actually begin until it has been provided. This is easy to do because
the download implementation is already internally asynchronous. A basic
test is added to make sure it even works.

This commit also contains a few drive-by changes in TestDownloads.cpp,
notably where two of the decide-destination callbacks were missing
return values and therefore triggering undefined behavior. It also fixes
a typo that got copy/pasted around and removes some references to URIs
that are now paths
* Source/WebKit/UIProcess/API/glib/WebKitDownload.cpp:
(_WebKitDownloadPrivate::~_WebKitDownloadPrivate):
(maybeFinishDecideDestination):
(webkitDownloadDecideDestination):
(webkit_download_class_init):
(webkitDownloadDecideDestinationWithSuggestedFilename):
(webkit_download_set_destination):
(webkit_download_cancel):
* Source/WebKit/UIProcess/API/glib/WebKitDownloadClient.cpp:
* Source/WebKit/UIProcess/API/glib/WebKitDownloadPrivate.h:
* Tools/TestWebKitAPI/Tests/WebKitGLib/TestDownloads.cpp:
(downloadLocalFileSuccessfully):
(testDownloadOverwriteDestinationDisallowed):
(testDownloadLocalFileError):
(testDownloadRemoteFile):
(testDownloadRemoteFileError):
(testDownloadAsyncDecideDestination):
(testDownloadAsyncDecideDestinationCancel):
(testDownloadMIMEType):
(testDownloadTextPlainMIMEType):
(testDownloadUserAgent):
(testDownloadEphemeralContext):
(testDownloadDestinationURI):
(beforeAll):

Canonical link: <a href="https://commits.webkit.org/261118@main">https://commits.webkit.org/261118@main</a>
</pre>

https://github.com/WebKit/WebKit/commit/449d64677d91fd566f5c9f42ad3a03e35422dfcc

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/6/builds/110584 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/77/builds/19664 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/43/builds/43175 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/87/builds/1937 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/12/builds/119449 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/11/builds/114546 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/76/builds/21086 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/85/builds/10793 "Built successfully") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/36/builds/102856 "Built successfully") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/19/builds/116326 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/78/builds/15704 "Passed tests") | [  ~~🧪 api-mac~~](https://ews-build.webkit.org/#/builders/3/builds/98895 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/35/builds/43982 "Passed tests") | 
| | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/9/builds/97664 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/73/builds/30562 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/34/builds/85892 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/81/builds/12316 "Built successfully") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/70/builds/31900 "Passed tests") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/82/builds/12876 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/84/builds/8857 "Passed tests") | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/80/builds/18240 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/62/builds/51516 "Passed tests") | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/79/builds/14760 "Built successfully") | | | 
| [❌ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/75/builds/4197 "Failed to push commit to Webkit repository") | | | |

### CODE REVIEW

**Comment by carlosgcampos** [comment]
I think it's better to split this. Since we are in a hurry to land these API changes I've moved the destination path part to https://github.com/WebKit/WebKit/pull/10855 using a simpler approach to avoid ifdefs.

File: Source/WebKit/UIProcess/API/glib/WebKitDownload.cpp
Comment (carlosgcampos) [inline Source/WebKit/UIProcess/API/glib/WebKitDownload.cpp]: CompletionHandler already asserts when destroyed without being invoked, I'm not sure we need this.

File: Source/WebKit/UIProcess/API/glib/WebKitDownload.cpp
Comment (carlosgcampos) [inline Source/WebKit/UIProcess/API/glib/WebKitDownload.cpp]: I don't think we need to use `std::optional` CompletionHandler can be set to nullptr

File: Source/WebKit/UIProcess/API/glib/WebKitDownload.cpp
Comment (carlosgcampos) [inline Source/WebKit/UIProcess/API/glib/WebKitDownload.cpp]: if (auto completionHandler = std::exchange(download->priv->decideDestinationCallback, nullptr)) {

File: Source/WebKit/UIProcess/API/glib/WebKitDownload.cpp
Comment (carlosgcampos) [inline Source/WebKit/UIProcess/API/glib/WebKitDownload.cpp]: String::fromUTF8 will return an empty string if destination is nullptr, I think. But this is not what we want, we say we use the suggested filename to sabe it under the downloads dir when not handled. If this was handled, then we should have a destination.

File: Source/WebKit/UIProcess/API/glib/WebKitDownload.cpp
Comment (mcatanzaro) [reply Source/WebKit/UIProcess/API/glib/WebKitDownload.cpp]: Well it's optional to check this here, but (a) a g_critical() matches the precondition assertions we do using g_return_if_fail(), and (b) CompletionHandler asserts are only enabled in debug builds anyway, so app developers will never see them.

File: Source/WebKit/UIProcess/API/glib/WebKitDownload.cpp
Comment (mcatanzaro) [reply Source/WebKit/UIProcess/API/glib/WebKitDownload.cpp]: Oops, didn't know.

File: Source/WebKit/UIProcess/API/glib/WebKitDownload.cpp
Comment (mcatanzaro) [reply Source/WebKit/UIProcess/API/glib/WebKitDownload.cpp]: Believe it or not, I actually wrote this at first because I've seen you do it several times and it seemed clever. Then I got rid of it to reduce horizontal scrolling. :P

File: Source/WebKit/UIProcess/API/glib/WebKitDownload.cpp
Comment (mcatanzaro) [reply Source/WebKit/UIProcess/API/glib/WebKitDownload.cpp]: Here we don't know whether it was handled or not. This code runs in both conditions.

File: Source/WebKit/UIProcess/API/glib/WebKitDownload.cpp
Comment (mcatanzaro) [reply Source/WebKit/UIProcess/API/glib/WebKitDownload.cpp]: Anyway, I'll change this as you suggest.

**Comment by mcatanzaro** [comment]
> I think it's better to split this. Since we are in a hurry to land these API changes I've moved the destination path part to https://github.com/WebKit/WebKit/pull/10855 using a simpler approach to avoid ifdefs.

OK, so what I'll do is (a) wait for #10855 to land, (b) also fix [this bug](https://bugs.webkit.org/show_bug.cgi?id=252827) and wait for it that to land too, then (c) update this pull request.

**Comment by carlosgcampos** [comment]
> > I think it's better to split this. Since we are in a hurry to land these API changes I've moved the destination path part to #10855 using a simpler approach to avoid ifdefs.
> 
> OK, so what I'll do is (a) wait for #10855 to land, (b) also fix [this bug](https://bugs.webkit.org/show_bug.cgi?id=252827) and wait for it that to land too, then (c) update this pull request.

I would move the final type thing to a different pr too.

**Comment by mcatanzaro** [comment]
> I would move the final type thing to a different pr too.

Yes, that's (b).

**Comment by mcatanzaro** [comment]
Ah, I found a flaw when porting Epiphany: we should also finish the callback if webkit_download_cancel() is called.

File: Tools/TestWebKitAPI/Tests/WebKitGLib/TestDownloads.cpp
Comment (carlosgcampos) [inline Tools/TestWebKitAPI/Tests/WebKitGLib/TestDownloads.cpp]: Why is this static? it uses m_suggstedFilename and m_download

File: Tools/TestWebKitAPI/Tests/WebKitGLib/TestDownloads.cpp
Comment (carlosgcampos) [inline Tools/TestWebKitAPI/Tests/WebKitGLib/TestDownloads.cpp]: Here you can do test->decideDestination()

File: Tools/TestWebKitAPI/Tests/WebKitGLib/TestDownloads.cpp
Comment (carlosgcampos) [inline Tools/TestWebKitAPI/Tests/WebKitGLib/TestDownloads.cpp]: And here too

**Review by carlosgcampos — APPROVED** [review]
Please add a test for cancel case too, it can be in a follow up if you prefer.

File: Tools/TestWebKitAPI/Tests/WebKitGLib/TestDownloads.cpp
Comment (mcatanzaro) [reply Tools/TestWebKitAPI/Tests/WebKitGLib/TestDownloads.cpp]: There is a virtual member function with the same name. Probably too confusing. I'll just get rid of it.

Comment (mcatanzaro) [reply Tools/TestWebKitAPI/Tests/WebKitGLib/TestDownloads.cpp]: Decided to rename it, actually.

**Comment by mcatanzaro** [comment]
> Please add a test for cancel case too, it can be in a follow up if you prefer.

I'm having some trouble writing the test, but that likely indicates that something's wrong, so I'll keep investigating.

**Comment by mcatanzaro** [comment]
OK, revised commit incoming. I'll set request-merge-queue to give you a chance to review the new test before it lands.

**Comment by mcatanzaro** [comment]
> OK, revised commit incoming. I'll set request-merge-queue to give you a chance to review the new test before it lands.

Done. Also there was no problem except with my ability to write the test. ;)

**Review by carlosgcampos — APPROVED** [review]

**Pull request merged**
